### PR TITLE
Add ability to add and retract statements by `sid`

### PR DESCRIPTION
**Pull request by kennyjwilli**
**Builds on https://github.com/luchiniatwork/saku/pull/10. We should merge that first.** 
---

Saku currently has the ability to upsert an identity or resource policy to a given DRN. The `upsert` will actually override the full `:policy/statements` list with the new policy statement list. In the below example we try adding a new policy that grants `user1` access to `drn11`. Instead of adding the new statement, all statements get replaced. 

```clojure
(dal/upsert-identity-policies dal-obj
  [{:policy/drn        "user1"
    :policy/statements [{:statement/actions   ["a1" "a2"]
                         :statement/effect    [:effect :allow]
                         :statement/resources ["drn1" "drn2"]}]}])
=>
[{:db/id 3,
  :policy/drn "user1",
  :policy/statements [{:db/id 4,
                       :statement/actions ["a1" "a2"],
                       :statement/effect {:db/id 2, :effect :allow},
                       :statement/resources ["drn1" "drn2"]}]}]

;; Do another upsert, adding `drn3` to the list of resources.
(dal/upsert-identity-policies dal-obj
  [{:policy/drn        "user1"
    :policy/statements [{:statement/actions   ["b1" ]
                         :statement/effect    [:effect :allow]
                         :statement/resources ["drn11"]}]}])
=>
[{:db/id 3,
  :policy/drn "user1",
  :policy/statements [{:db/id 5,
                       :statement/actions ["b1"],
                       :statement/effect {:db/id 2, :effect :allow},
                       :statement/resources ["drn11"]}]}]
```

While initially intuitive, this behavior results in racy code when trying to add or remove statements from a identity or policy. e.g., you’d first need to get the identity policies, modify then, and then upsert.

To address this, we add the ability to add statements by a statement id `sid`. This allows users to ensure a policy with a given `sid` is added to a resource or identity and to remove exactly that policy when desired.

### CODE REVIEW

File: modules/saku-policy-store/src/saku/dal_datalevin.clj [hunk 1]
```diff
@@ -52,38 +43,77 @@
...
+
+(defn tx-fn-retract-statements-from-policy
+  [db {:keys [policy/drn statement-ids]}]
+  (let [eids-to-retract (d/q '[:find ?s
+                               :in $ ?policy [?statement-id ...]
+                               :where
+                               [?policy :policy/statements ?s]
+                               [?s :statement/sid ?statement-id]]
+                          db [:policy/drn drn] statement-ids)]
+    (map (fn [[eid]] [:db/retractEntity eid]) eids-to-retract)))
```
Comment (kennyjwilli) [inline modules/saku-policy-store/src/saku/dal_datalevin.clj:91]: We use a transaction function to implement the retract for a given policy drn + sid. The other option was to add a unique identifier like `policy-drn+sid` to every statement. This felt like overkill, and I didn't like the duplication. Fortunately datalevin has great, native support for transaction functions, so we can implement this natively without any model changes!

File: modules/saku-policy-store/src/saku/dal_datalevin.clj [hunk 2]
```diff
@@ -52,38 +43,77 @@
...
+
+(defn add-statements*
+  [db-conn statements-input]
+  (let [{:keys [policy/drn policy/statements]} statements-input
+        retract-sids (into [] (keep :statement/sid) statements)
+        retract-tx (when (seq retract-sids)
+                     [:db.fn/call tx-fn-retract-statements-from-policy
+                      {:policy/drn    drn
+                       :statement-ids retract-sids}])
```
Comment (kennyjwilli) [inline modules/saku-policy-store/src/saku/dal_datalevin.clj:100]: We only need to retract statements that have a `sid`, else we assume the user intends to add an unidentified statement to the policy.

**Review by kennyjwilli — COMMENTED** [review]
author's review

**Comment by luchiniatwork** [comment]
Rebased this one manually to manage confluicts.

**Pull request closed without merging**